docs/source/rfc/isa_restructuring.org

```diff
@@ -0,0 +1,732 @@
+
+The purpose of this document is thus to propose some much needed
+change to the current Complex Instruction Set Computer (architecture)
+based on both a theoretical argument for the scalability of RISC, and
+some practical observations that have been made over the span of
+2021-2023 about the existing and projected uses for the Iroha
+Distributed Ledger Technology. As such it should be treated as a mere
+outline rather than a precise specification.
+
+* Relation to the Executor model
+
+The proposed changes to the Instruction set architecture depend on
+prior art, titled Executor RFC. For some background, Iroha is intended
+as both a ledger for private (permissioned) and public (mistakenly
+known as permissionless) blockchains. The intent was that there would
+be a stack-based virtual machine that would operate on a set of Iroha
+Special Instructions, which shall be shared between the two blockchain
+types, but operate under different semantics, which are chosen at
+compilation time. This was later proven to be insufficient, because
+while almost all public blockchains operate under the same set of
+permissions, private blockchains often not only differ between each
+other, but their permission models also frequently require adjustment
+post-deployment.
+
+Under these circumstances it was decided to make the permission
+validator a dynamic module that is loaded at runtime and is subject to
+upgrade provided that the requisite conditions specified inside the
+module itself are met (e.g. governance vote).  This in turn exposed
+several inefficiencies and some tension between what is considered a
+blockchain invariant and what comprises a blockchain-agnostic,
+universal invariant that must be upheld on all Iroha-based
+blockchains.  To address those inefficiencies it was decided that
+instead of a dynamically loadable permission validator, there would be
+a dynamically loadable executor, which is responsible not only for
+checking if the instruction application violates any of the
+invariants, and checks for permissions, but is responsible wholly for
+upholding the blockchain invariants.
+
+As such, under the executor model, instructions are exposed as objects
+for which the application is a well-defined fallible operation.  The
+executor operates on a per-transaction basis, meaning that if a set of
+instructions is submitted, in principle, the executor can allow or
+deny the applications of certain instruction combinations.  At which
+point the permission tokens became ill-defined, as their meaning is
+undefined until a specific executor is loaded.  Within these
+constraints, the distinction between blockchain and host invariants
+becomes clearly defined: everything that isn't enforced by a failure
+in applying an instruction is a blockchain invariant, including
+permissions. The executor is responsible for enforcing the blockchain
+invariants, and enforcement of host instructions is propagated from
+the host to the executor as well.
```

Comment:
> and enforcement of host instructions is propagated from the host to the executor as well

What does it mean, I don't get it

Reply:
Roughly, that the host no longer checks if an instruction violates any invariants. There's no distinction between host invariants and blockchain invariants anymore and so the logic that says "Instruction X bad, because it caused overflow" is moved up into the executor as a static library, not as a virtual call of a host function.

Reply:
> At which point the permission tokens became ill-defined, as their meaning is undefined until a specific executor is loaded

I don't agree with this point. Permission tokens are not ill-defined until executor is loaded, rather they don't exist at all (i.e. permission tokens set is empty and you can verify this with `FindPermissionTokens` query)